In [1]:
import pandas as pd 
import numpy as np

In [18]:
# Task 1 — Fix Hidden Missing Values
df = pd.read_csv('E:/Data Analysis learning/labs/adult.csv')
df.head()


,age,workclass,fnlwgt,education,education.num,marital.status,occupation,relationship,race,sex,capital.gain,capital.loss,hours.per.week,native.country,income
0,90,?,77053,HS-grad,9,Widowed,?,Not-in-family,White,Female,0,4356,40,United-States,<=50K
1,82,Private,132870,HS-grad,9,Widowed,Exec-managerial,Not-in-family,White,Female,0,4356,18,United-States,<=50K
2,66,?,186061,Some-college,10,Widowed,?,Unmarried,Black,Female,0,4356,40,United-States,<=50K
3,54,Private,140359,7th-8th,4,Divorced,Machine-op-inspct,Unmarried,White,Female,0,3900,40,United-States,<=50K
4,41,Private,264663,Some-college,10,Separated,Prof-specialty,Own-child,White,Female,0,3900,40,United-States,<=50K


In [73]:
df.shape

(32536, 15)

In [19]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32561 entries, 0 to 32560
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   age             32561 non-null  int64 
 1   workclass       32561 non-null  object
 2   fnlwgt          32561 non-null  int64 
 3   education       32561 non-null  object
 4   education.num   32561 non-null  int64 
 5   marital.status  32561 non-null  object
 6   occupation      32561 non-null  object
 7   relationship    32561 non-null  object
 8   race            32561 non-null  object
 9   sex             32561 non-null  object
 10  capital.gain    32561 non-null  int64 
 11  capital.loss    32561 non-null  int64 
 12  hours.per.week  32561 non-null  int64 
 13  native.country  32561 non-null  object
 14  income          32561 non-null  object
dtypes: int64(6), object(9)
memory usage: 3.7+ MB


In [20]:
df.replace('?',np.nan,inplace= True)

In [34]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 32561 entries, 0 to 32560
Data columns (total 15 columns):
 #   Column          Non-Null Count  Dtype 
---  ------          --------------  ----- 
 0   age             32561 non-null  int64 
 1   workclass       30725 non-null  object
 2   fnlwgt          32561 non-null  int64 
 3   education       32561 non-null  object
 4   education.num   32561 non-null  int64 
 5   marital.status  32561 non-null  object
 6   occupation      30718 non-null  object
 7   relationship    32561 non-null  object
 8   race            32561 non-null  object
 9   sex             32561 non-null  object
 10  capital.gain    32561 non-null  int64 
 11  capital.loss    32561 non-null  int64 
 12  hours.per.week  32561 non-null  int64 
 13  native.country  31978 non-null  object
 14  income          32561 non-null  object
dtypes: int64(6), object(9)
memory usage: 3.7+ MB


In [29]:
# Quantify missingness
df.isnull().sum().sort_values(ascending=False)

occupation        1843
workclass         1836
native.country     583
fnlwgt               0
education            0
education.num        0
age                  0
marital.status       0
relationship         0
sex                  0
race                 0
capital.gain         0
capital.loss         0
hours.per.week       0
income               0
dtype: int64

In [30]:
(df.isnull().mean()*100).sort_values(ascending= False)

occupation        5.660146
workclass         5.638647
native.country    1.790486
fnlwgt            0.000000
education         0.000000
education.num     0.000000
age               0.000000
marital.status    0.000000
relationship      0.000000
sex               0.000000
race              0.000000
capital.gain      0.000000
capital.loss      0.000000
hours.per.week    0.000000
income            0.000000
dtype: float64

In [41]:
df['native.country'].value_counts(dropna= False)

native.country
United-States                 29170
Mexico                          643
NaN                             583
Philippines                     198
Germany                         137
Canada                          121
Puerto-Rico                     114
El-Salvador                     106
India                           100
Cuba                             95
England                          90
Jamaica                          81
South                            80
China                            75
Italy                            73
Dominican-Republic               70
Vietnam                          67
Guatemala                        64
Japan                            62
Poland                           60
Columbia                         59
Taiwan                           51
Haiti                            44
Iran                             43
Portugal                         37
Nicaragua                        34
Peru                             31
Greece       

In [40]:
# Task 3 — Handle Missing Values
df['workclass'].fillna('unknown',inplace=True)
df['occupation'].fillna('unknown',inplace=True)

In [42]:
df['native.country'].fillna('unknown',inplace=True)

In [43]:
# Task 4 — Detect Duplicates
df.duplicated().sum()


np.int64(24)

In [45]:
cols = ['age', 'workclass', 'fnlwgt', 'education', 'education.num',
       'marital.status', 'occupation', 'relationship', 'race', 'sex',
       'capital.gain', 'capital.loss', 'hours.per.week', 'native.country']
df[cols].duplicated().sum()

np.int64(25)

In [ ]:
# yes people having identical attributes and different outcomes is considered as duplicate

In [47]:
df.drop_duplicates(inplace=True)

In [51]:
df.drop_duplicates(subset=cols,inplace=True)

In [53]:
# Task 5 Fix Inconsistent Categorical Data

In [54]:
df['education'].unique()

array(['HS-grad', 'Some-college', '7th-8th', '10th', 'Doctorate',
       'Prof-school', 'Bachelors', 'Masters', '11th', 'Assoc-acdm',
       'Assoc-voc', '1st-4th', '5th-6th', '12th', '9th', 'Preschool'],
      dtype=object)

In [55]:
df['marital.status'].unique()

array(['Widowed', 'Divorced', 'Separated', 'Never-married',
       'Married-civ-spouse', 'Married-spouse-absent', 'Married-AF-spouse'],
      dtype=object)

In [56]:
df['native.country'].unique()

array(['United-States', 'unknown', 'Mexico', 'Greece', 'Vietnam', 'China',
       'Taiwan', 'India', 'Philippines', 'Trinadad&Tobago', 'Canada',
       'South', 'Holand-Netherlands', 'Puerto-Rico', 'Poland', 'Iran',
       'England', 'Germany', 'Italy', 'Japan', 'Hong', 'Honduras', 'Cuba',
       'Ireland', 'Cambodia', 'Peru', 'Nicaragua', 'Dominican-Republic',
       'Haiti', 'El-Salvador', 'Hungary', 'Columbia', 'Guatemala',
       'Jamaica', 'Ecuador', 'France', 'Yugoslavia', 'Scotland',
       'Portugal', 'Laos', 'Thailand', 'Outlying-US(Guam-USVI-etc)'],
      dtype=object)

In [ ]:
# Task 6 — Descriptive Statistics

In [57]:
df.describe(include = [np.number])

# mean age = 38 median age = 37 they don't differ significantly
# mean hours_per_week = 40 median = 49 alomost equal  

,age,fnlwgt,education.num,capital.gain,capital.loss,hours.per.week
count,32536.000000,3.253600e+04,32536.000000,32536.000000,32536.000000,32536.000000
mean,38.585536,1.897824e+05,10.081725,1078.476887,87.370912,40.440343
std,13.638193,1.055577e+05,2.571622,7388.068543,403.107737,12.347079
min,17.000000,1.228500e+04,1.000000,0.000000,0.000000,1.000000
25%,28.000000,1.178242e+05,9.000000,0.000000,0.000000,40.000000
50%,37.000000,1.783560e+05,10.000000,0.000000,0.000000,40.000000
75%,48.000000,2.369932e+05,12.000000,0.000000,0.000000,45.000000
max,90.000000,1.484705e+06,16.000000,99999.000000,4356.000000,99.000000


In [58]:
df.describe(include = [object])

,workclass,education,marital.status,occupation,relationship,race,sex,native.country,income
count,32536,32536,32536,32536,32536,32536,32536,32536,32536
unique,9,16,7,15,6,5,2,42,2
top,Private,HS-grad,Married-civ-spouse,Prof-specialty,Husband,White,Male,United-States,<=50K
freq,22672,10494,14969,4136,13186,27794,21774,29152,24698


In [86]:
df[''].value_counts()

race
White                 27794
Black                  3122
Asian-Pac-Islander     1038
Amer-Indian-Eskimo      311
Other                   271
Name: count, dtype: int64

In [63]:
#Task 7— Frequency Analysis
df['occupation'].value_counts().sort_values(ascending=False).head(1)

occupation
Prof-specialty    4136
Name: count, dtype: int64

In [65]:
(df['sex'].value_counts(normalize=True)*100).sort_values(ascending=False)

sex
Male      66.922793
Female    33.077207
Name: proportion, dtype: float64

In [66]:
(df['income'].value_counts(normalize=True)*100).sort_values(ascending=False)

income
<=50K    75.909761
>50K     24.090239
Name: proportion, dtype: float64

In [ ]:
# incomes are not balanced

In [ ]:
# Task 8— Cross-Check a Data Quality Assumption

In [67]:
df.head()

,age,workclass,fnlwgt,education,education.num,marital.status,occupation,relationship,race,sex,capital.gain,capital.loss,hours.per.week,native.country,income
0,90,unknown,77053,HS-grad,9,Widowed,unknown,Not-in-family,White,Female,0,4356,40,United-States,<=50K
1,82,Private,132870,HS-grad,9,Widowed,Exec-managerial,Not-in-family,White,Female,0,4356,18,United-States,<=50K
2,66,unknown,186061,Some-college,10,Widowed,unknown,Unmarried,Black,Female,0,4356,40,United-States,<=50K
3,54,Private,140359,7th-8th,4,Divorced,Machine-op-inspct,Unmarried,White,Female,0,3900,40,United-States,<=50K
4,41,Private,264663,Some-college,10,Separated,Prof-specialty,Own-child,White,Female,0,3900,40,United-States,<=50K


In [71]:
df.groupby('education')['education.num'].unique()

education
10th             [6]
11th             [7]
12th             [8]
1st-4th          [2]
5th-6th          [3]
7th-8th          [4]
9th              [5]
Assoc-acdm      [12]
Assoc-voc       [11]
Bachelors       [13]
Doctorate       [16]
HS-grad          [9]
Masters         [14]
Preschool        [1]
Prof-school     [15]
Some-college    [10]
Name: education.num, dtype: object

In [82]:
import pandas as pd
import sqlite3

# 2. Connect to SQLite database
conn = sqlite3.connect("adult.db")

# 3. Upload DataFrame to SQLite table
df.to_sql(
    "adult",
    conn,
    if_exists="replace",
    index=False
)

# 4. Close connection
conn.close()

print("Data uploaded successfully!")

Data uploaded successfully!


In [84]:


conn = sqlite3.connect("adult.db")

df = pd.read_sql("SELECT * FROM adult", conn)

print(df)

conn.close()

       age workclass  fnlwgt     education  education.num      marital.status  \
0       90   unknown   77053       HS-grad              9             Widowed   
1       82   Private  132870       HS-grad              9             Widowed   
2       66   unknown  186061  Some-college             10             Widowed   
3       54   Private  140359       7th-8th              4            Divorced   
4       41   Private  264663  Some-college             10           Separated   
...    ...       ...     ...           ...            ...                 ...   
32531   22   Private  310152  Some-college             10       Never-married   
32532   27   Private  257302    Assoc-acdm             12  Married-civ-spouse   
32533   40   Private  154374       HS-grad              9  Married-civ-spouse   
32534   58   Private  151910       HS-grad              9             Widowed   
32535   22   Private  201490       HS-grad              9       Never-married   

              occupation   

# Task 10— Data Profiling Summary Report
### 1. Dataset Overview
rows and cols = (32536, 15)
It represent attributes of people and their income (less or more than 50 or)
### 2. Data quality issues found
occupation,workclass and country has missing values. 25 duplicate values found, no incosistent data found
### 3. Key observations
Male has 66% whereas female has only 33% of our data. 75 % people has less than 50k indicating higher % have low income
The most common occupation is Prof-specialty
### 4. Cleaning decisions made
I replace null values with unknown for categorical column. dropped duplicated rows.


In [81]:
print(f"Driver={{SQL Server}};"
f"Server={server};"
f"Database={database};"
f"Trusted_Connection=yes;")

Driver={SQL Server};Server=localhost/SQLEXPRESS;Database=labs;Trusted_Connection=yes;
